src/diagnostic-channel-publishers/src/postgres.pub.ts

```diff
@@ -0,0 +1,128 @@
+            return channel.bindToContext(function trackingCallback(err: Error, res: IPostgresResult): any {
+                const end = process.hrtime(start);
+                data.result = res && { rowCount: res.rowCount, command: res.command };
+                data.error = err;
+                data.duration = Math.ceil((end[0] * 1e3) + (end[1] / 1e6));
+                channel.publish("postgres", data);
+
+                // emulate weird internal behavior in pg@6
+                // on success, the callback is called *before* query events are emitted
+                // on failure, the callback is called *instead of* the query emitting events
+                // with no events, that means no promises (since the promise is resolved/rejected in a )
+                // since we are always inserting ourselves as a callback, we have to restore the original
+                // behavior if the user didn't provide one themselves
+                if (err) {
+                    if (cb) {
+                        return cb.apply(this, arguments);
+                    } else if (queryResult && queryResult instanceof EventEmitter) {
+                        queryResult.emit("error", err);
+                    }
+                } else if (cb) {
+                    cb.apply(this, arguments);
+                }
+            });
+        }
+
+        // this function takes too many variations of aguments.
+        // this patches any provided callback or creates a new callback if one wasn't provided.
+        // since the callback is always called (if provided) in addition to always having a Promisified
+        // EventEmitter returned (well, sometimes -- see above), its safe to insert a callback if none was given
+        if (typeof config === "string") {
+            if (values instanceof Array) {
+                data.query.preparable = {
+                    text: config,
+                    args: values,
+                };
+                callback = patchCallback(callback);
+            } else {
+                data.query.text = config;
+                values = patchCallback(values);
+            }
+        } else {
+            if (typeof config.name === "string") {
+                data.query.plan = config.name;
+            } else if (config.values instanceof Array) {
+                data.query.preparable = {
+                    text: config.text,
+                    args: config.values,
+                };
+            } else {
+                data.query.text = config.text;
+            }
+
+            if (config.callback) {
+                config.callback = patchCallback(config.callback);
+            } else if (values) {
+                values = patchCallback(values);
+            } else {
+                // user didn't provide a callback
+                values = patchCallback();
+            }
+        }
+
+        queryResult = originalClientQuery.call(this, config, values, callback);
```

Comment:
The original module never uses arguments internally, but I also assumed that reassigning config/values/callback wouldn't update the arguments array, but I just checked and it does. I can fix this.